R/ComputeShifts.R

```diff
@@ -0,0 +1,100 @@
+#' Adds leads or lags of a variable, robustly to gaps in time variable, as new columns in a panel dataset
+#'
+#' @param df Data frame that will be modified.
+#' @param idvar Character indicating column of units.
+#' @param timevar Character indicating column of time periods.
+#' @param shiftvar Character indicating column of variable that will be shifted.
+#' @param shiftvalues Numeric vector specifying the leads/lags to be computed. For example, '-1:1' will compute the lag, current, and lead values.
+#' @param timevar_holes Logical indicating whether the panel contains gaps in the time variable.
+#' @param return_df Logical indicating whether the function should return a data frame.
+#'
+#' @return The passed dataset augmented with columns that reflect the desired shift values
+
+#' @seealso [data.table::shift()]
+#'
+#' @examples
+#' ComputeShifts(
+#'     example_data,
+#'     idvar = "id",
+#'     timevar = "t",
+#'     shiftvar = "z",
+#'     shiftvalues = -2:2
+#' )
+#'
+#'
+#' @importFrom data.table setDT setorderv shift := CJ .SD
+#' @keywords internal
+#' @noRd
+
+ComputeShifts <- function(df, idvar, timevar, shiftvar, shiftvalues,
+                          timevar_holes = FALSE, return_df = TRUE) {
+    if (! is.data.frame(df)) {
+        stop("df should be a data frame.")
+    }
+    for (var in c(idvar, timevar, shiftvar)) {
+        if ((! is.character(var))) {
+            stop(paste0(var, " should be a character."))
+        }
+        if (! var %in% colnames(df)) {
+            stop(paste0(var, " should be the name of a variable in the dataset."))
+        }
+    }
+    if (! is.logical(timevar_holes)) {
+        stop("timevar_holes should be logical.")
+    }
+    if (! is.logical(return_df)) {
+        stop("return_df should be logical.")
+    }
+    if (! is.numeric(shiftvalues)) {
+        stop("shiftvalues should be numeric.")
+    }
+    if (0 %in% shiftvalues & length(shiftvalues) == 1) {
+        stop("shiftvalues must be different than 0 to compute leads/lags.")
+    }
+
+    data.table::setDT(df)
+    data.table::setorderv(df, cols = c(idvar, timevar))
+
+    lags  <-    shiftvalues[shiftvalues >  0]
+    leads <- -1*shiftvalues[shiftvalues <= 0]
+
+    compute_shifts <- function(dat, idvar, shiftvar, shiftvalues, lags, leads) {
+
+        if (any(shiftvalues > 0)) {
+            dat[, paste0(shiftvar, "_lag", lags) :=
+                     data.table::shift(get(shiftvar), lags, type = "lag"),
+                 by = c(idvar)]
+        }
+        if (any(shiftvalues <= 0)) {
+            dat[, paste0(shiftvar, "_lead", leads) :=
+                     data.table::shift(get(shiftvar), leads, type = "lead"),
+                 by = c(idvar)]
+        }
+        return(dat)
+    }
+
+    if (!timevar_holes) {
```

Comment:
The code to find holes/gaps in the time variables is in `EventStudy`. We then pass a boolean to `ComputeFirstDifferences` and `ComputeShifts`. 

https://github.com/JMSLab/eventstudyr/blob/8dd8d5d1bf87272bf4b7d1fff7e93b7619f92009/R/EventStudy.R#L182-L199

What we do in each function when `timevar_holes` is TRUE is slightly different, so not sure adding a helper function here would help.

I agree in that it would be useful to have a separate function to find the gaps @nateschor. This is the second bullet-point under "Short-term" in the [Roadmap](https://github.com/JMSLab/eventstudyr/wiki/Roadmap).

Reply:
Since the `timevar_holes` functions are slightly different, I agree that a helper function wouldn't be worthwhile at the moment.

Adding to the Roadmap a function for finding the gaps sounds good, thanks @santiagohermo!